### PR TITLE
Prepare types

### PR DESCRIPTION
**Pull request by tom-wa**
# Purpose

prepare type plugins for type dispatcher

# Checklist

- [x] commit messages are fine (with references to issues)
- [x] I ran all tests and everything went fine
- [ ] I added unit tests
- [ ] affected documentation is fixed
- [ ] I added code comments, logging, and assertions
- [ ] meta data is updated (e.g. README.md of plugins)

### CODE REVIEW

File: src/plugins/path/path.c [hunk 1]
```diff
@@ -12,6 +12,60 @@
 #include "kdbconfig.h"
 #endif
 
+static int validateKey(Key *key, Key *parentKey)
```
Comment (markus2330) [inline src/plugins/path/path.c]: Seems like code is not formatted yet. And parentKey should be renamed to errorKey?

Don't you want to export this function? (Why is it static?)

**Comment by tom-wa** [comment]
yeah, i'm still working on it because the current signature doesn't really work for the boolean checks.

**Comment by markus2330** [comment]
Please try to keep the interface as simple as possible. For boolean: Is canonicalization and the validation separated? Why isn't the interface enough?

**Comment by tom-wa** [comment]
boolean has some hardcoded default values + to plugin configuration

**Comment by tom-wa** [comment]
how should keys that fail to validate on kdbGet be handled ? 
should kdbGet fail with an error, or just delete all keys that fail validation (+ maybe a warning) ?

**Comment by markus2330** [comment]
> how should keys that fail to validate on kdbGet be handled ? should kdbGet fail with an error, or just delete all keys that fail validation (+ maybe a warning) ?

Remove them **with** a warning and let us write a plugin that allows us to turn this warning into an error. (For applications that want to fail.) For admins and so on there should always be a way with Elektra to fix the situation, and if `kdbGet` would fail, they could not fix it. See #1291

**Comment by tom-wa** [comment]
should the validation plugins remove the key or only the type dispatcher ? 
does that mean validation plugins should only set warnings but never errors ?

**Comment by markus2330** [comment]
> should the validation plugins remove the key or only the type dispatcher ?

I would say the type dispatcher. Maybe we can even configure the type dispatcher for different behaviour?

> does that mean validation plugins should only set warnings but never errors ?

On writing configuration they certainly set errors and fail!

**Comment by tom-wa** [comment]
>I would say the type dispatcher. Maybe we can even configure the type dispatcher for different behaviour?

if it's done by the type dispatcher we can use `validateX(Key * key, Key * errorKey)` as signature. The validation plugins would use the parentKey as errorKey, the type dispatcher a new key and then sets a warning with the description/text of the error. 

what do you think about generating both SET_ERROR and ADD_WARNING macros for every entry in `error/specification` ?

**Comment by markus2330** [comment]
> if it's done by the type dispatcher we can use validateX(Key * key, Key * errorKey) as signature.

Sounds good!

> The validation plugins would use the parentKey as errorKey, the type dispatcher a new key and then sets a warning with the description/text of the error.

Okay!

> what do you think about generating both SET_ERROR and ADD_WARNING macros for every entry in error/specification ?

I am not in favour of it. It should be a rare exception that errors/warnings are up/downgraded.

Maybe you are right and we should let the validation plugins do nothing in `kdbGet` (neither iterate, nor check, no warning). Then only the type dispatcher allows us to validate within `kdbGet`, which should be okay.

We need an architecture document for these decision, can you write such a decision about validation? (With complete rationale and all pro/cons)

File: src/plugins/date/date.c [hunk 2]
```diff
@@ -465,6 +465,7 @@ int elektraDateGet (Plugin * handle ELEKTRA_UNUSED, KeySet * returned ELEKTRA_UN
 			       keyNew ("system/elektra/modules/date/exports", KEY_END),
 			       keyNew ("system/elektra/modules/date/exports/get", KEY_FUNC, elektraDateGet, KEY_END),
 			       keyNew ("system/elektra/modules/date/exports/set", KEY_FUNC, elektraDateSet, KEY_END),
+			       keyNew ("system/elektra/modules/date/export/validateKey", KEY_FUNC, validateKey, KEY_END),
```
Comment (markus2330) [inline src/plugins/date/date.c]: Its a Elektra key, we should avoid camelcase here. So either `validate` (should be enough?) or `validate/key` (if we might need further validation methods but seems unlikely: validating of keyset is already `kdbSet`). And there is a typo: `export` -> `exports`.

**Comment by tom-wa** [comment]
@markus2330 i have no idea how to export the validation function from the types plugin, could you do that ?

**Comment by markus2330** [comment]
Yes, should be easy.

Basically we only need to wrap `bool check (Key & k)`.
Can you please document the API of the validation function? In particular:

- signature
- will open/close be called as expected?
- Are there some restrictions which errors needs to be set?
- return values?

**Comment by tom-wa** [comment]
`int validateKey(const Key *key, const Key *errorKey)`
returns 1 if validation succeeded, 0 if validation failed
not really sure if an additional return value for invalid syntax is needed.
if an error occurred it's set on errorKey

open/close is not called, is it needed somewhere ?

**Comment by markus2330** [comment]
In the current version of the plugin `type`, `open` builds up a data structure (containing all types) which is used during checking keys. I could build up the data structure in the wrapper as well but I think it is somewhat expected that a plugin gets opened before anything is done with it? Would it be a problem to introduce this behavior?

And can you please add the docu within the source, please?

And we should follow the plugin convention and fail with -1? Or were you thinking about booleans as return value?

**Comment by tom-wa** [comment]
i forgot that elektraPluginOpen and elektraPluginClose call open/close by default, so open/close do get called.

initially i wanted to use 1 for success, 0 for error and -1 for syntax errors in the check/X metakeys. but i'm not so sure about the last party anymore. what do you think ?

**Comment by markus2330** [comment]
> initially i wanted to use 1 for success, 0 for error and -1 for syntax errors in the check/X metakeys. but i'm not so sure about the last party anymore. what do you think ?

I also was thinking about some intermediate state (like "could be okay, depends on other keys"), but it seems like that we do not need it. A simpler design is better, so lets stay with a boolean as return value.

**Comment by tom-wa** [comment]
now that you mention "depends on other keys" i realized i've forgotten about `conditionals` and `math` which both need the whole keyset + a parentkey if keynames relative to the parentkey are used. so `int validateKey(const Key *key, const Key *errorKey)` won't work for them. 
it should be `int validateKey( const Key *key, KeySet *ks, Key *errorKey)`

**Comment by markus2330** [comment]
The question is if the do not better fit into the transformation plugin (especially with the assignment feature).

I think it would be better if you wrap up the type checker for now. Let us merge it, test it a bit, and progress from there.

So please fix docu, comments and so on and set the PRs to ready to merge.

**Comment by markus2330** [comment]
The type plugin now exports validateKey, too.

Would be great to get the validation for lcdproc merged, soon!

**Comment by markus2330** [comment]
Thank you, great work!

**Pull request merged**